### PR TITLE
This change introduces using cl_arm_printf with callbacks for using printf inside the kernel

### PR DESCRIPTION
**Pull request by Rekt3421**
This changes is a part of a series which will address #578 . This change addresses point 3 out of 4.

### CODE REVIEW

File: tests/api/printf.cpp [hunk 1]
```diff
@@ -40,6 +40,15 @@ static void releaseStdout(int fd) {
     close(fd);
 }
 
+#define RETURN_ON_FAILURE(err, ret)                                            \
```
Comment (Rekt3421) [inline tests/api/printf.cpp:43]: This is similar to the main one. Was having some trouble using ASSERT_CL_SUCCESS in some cases so just copied and reused this.

**Comment by Rekt3421** [comment]
There is one slight issue though how do I verify the contents of the printf buffer? I could not access the string from within the callback ( I am assuming perhaps this is due to some synchronization issue?)

File: tests/api/printf.cpp [hunk 2]
```diff
@@ -104,64 +115,131 @@ static std::string getStdoutFileName(temp_folder_deletion& temp) {
...
+                     void* user_data) {
+    captured_output.append(buffer, len);
+}
+
+void get_command_queue_printf(cl_command_queue& queue, cl_context new_ctx,
+                              cl_device_id device) {
+    cl_int err;
+    queue = clCreateCommandQueue(new_ctx, device, 0, &err);
+    ASSERT_CL_SUCCESS(err);
+}
```
Comment (rjodinchr) [inline tests/api/printf.cpp:174]: This function feels unnecessary

**Review by rjodinchr — CHANGES_REQUESTED** [review]
You should rely on the framework used by `api_tests`. Everything is in `testcl.hpp`.
It will allow to easily manage release/destruction as well as testing the error code (removing the need for `RETURN_ON_FAILURE` that you added)

You will need to add things there to create your context I believe.

In the end, we should not have any function starting with `cl` in `printf.cpp`.

File: tests/api/printf.cpp [hunk 3]
```diff
@@ -104,64 +115,131 @@ static std::string getStdoutFileName(temp_folder_deletion& temp) {
...
+    auto device = get_device_test(platform);
+    cl_context_properties properties[] = {
+        /* Enable a printf callback function for this context. */
+        CL_PRINTF_CALLBACK_ARM, (cl_context_properties)printf_callback,
+
+        /* Request a minimum printf buffer size of 4MiB for devices in the
+        context that support this extension. */
+        CL_PRINTF_BUFFERSIZE_ARM, (cl_context_properties)0x100000,
+
+        CL_CONTEXT_PLATFORM, (cl_context_properties)gPlatform, 0};
```
Comment (rjodinchr) [inline tests/api/printf.cpp:188]: I understand that this is part of the sample example [here](https://registry.khronos.org/OpenCL/extensions/arm/cl_arm_printf.html), but I don't think it is needed.

**Pull request closed without merging**